toolkit/offchain/src/csl.rs

```diff
@@ -278,22 +307,65 @@ impl TransactionBuilderExt for TransactionBuilder {
 		Ok(())
 	}
 
-	fn set_required_fields_and_build(
+	fn balance_update_and_build(
 		&mut self,
 		ctx: &TransactionContext,
 	) -> Result<Transaction, JsError> {
-		self.add_collateral_inputs(ctx)?;
-		self.calc_script_data_hash(&convert_cost_models(
-			&ctx.protocol_parameters.plutus_cost_models,
-		))?;
-		self.add_required_signer(&ctx.payment_key_hash);
-		self.add_inputs_from_and_change_with_collateral_return(
-			&ctx.payment_utxos.to_csl()?,
-			cardano_serialization_lib::CoinSelectionStrategyCIP2::LargestFirstMultiAsset,
-			&ChangeConfig::new(&key_hash_address(&ctx.payment_key_hash, ctx.network)),
-			&ctx.protocol_parameters.collateral_percentage.into(),
-		)?;
-		self.build_tx()
+		// Tries if the largest UTXO is enough to cover collateral, if not, adds more UTXOs
+		// starting from the largest remaining.
```

Comment:
Maybe it should try without any collateral input for transactions that do not use scripts?

Reply:
I think so, yes. I think the script cache setup transaction didn't need one for example.

Reply:
It ended up with `if` and trying to balance txs without smart-contracts involved.